src/overlays/actors/ovl_En_Test4/z_en_test4.c

```diff
@@ -33,253 +35,288 @@ ActorInit En_Test4_InitVars = {
 };
 
 static s32 sIsLoaded = false;
+
 // "Night of ..."
-static s16 sNightMessages1[] = { 0x1BB4, 0x1BB5, 0x1BB6 };
+static s16 sNightOfTextIds2[] = { 0x1BB4, 0x1BB5, 0x1BB6 };
 // "Dawn of ..." (Note: first two message are the same)
-static s16 sDayMessages1[] = { 0x1BB2, 0x1BB2, 0x1BB3 };
+static s16 sDawnOfTextIds2[] = { 0x1BB2, 0x1BB2, 0x1BB3 };
 // "Night of ..."
-static s16 sNightMessages2[] = { 0x1BB4, 0x1BB5, 0x1BB6 };
+static s16 sNightOfTextIds[] = { 0x1BB4, 0x1BB5, 0x1BB6 };
 // "Dawn of ..." (Note: first two message are the same)
-static s16 sDayMessages2[] = { 0x1BB2, 0x1BB2, 0x1BB3 };
-static u16 D_80A43364[] = { CLOCK_TIME(6, 0), CLOCK_TIME(18, 0) };
+static s16 sDawnOfTextIds[] = { 0x1BB2, 0x1BB2, 0x1BB3 };
+
+static u16 sDayNightTransitionTimes[THREEDAY_DAYTIME_MAX] = {
+    CLOCK_TIME(6, 0),  // THREEDAY_DAYTIME_NIGHT
+    CLOCK_TIME(18, 0), // THREEDAY_DAYTIME_DAY
+};
 
-static s16 sCsIdList[2];
+static s16 sCsIdList[THREEDAY_DAYTIME_MAX];
 static s16 sCurCsId;
 
-void func_80A41D70(EnTest4* this, PlayState* play) {
-    if (this->csIdIndex != 0) {
-        Message_DisplaySceneTitleCard(play, sNightMessages1[CURRENT_DAY - 1]);
-    } else if ((sCsIdList[this->csIdIndex] < 0) || (play->actorCtx.flags & ACTORCTX_FLAG_1)) {
-        if (play->actorCtx.flags & ACTORCTX_FLAG_1) {
+/**
+ * Handles the transition from day-night and night-day.
+ * This does not handle DayTelop transitions.
+ * Only differs from `EnTest4_HandleDayNightSwap` with an extra telescope check when turning day without a cutscene.
+ */
+void EnTest4_HandleDayNightSwapFromInit(EnTest4* this, PlayState* play) {
+    if (this->daytimeIndex != THREEDAY_DAYTIME_NIGHT) {
+        // Previously day, turning night
+        Message_DisplaySceneTitleCard(play, sNightOfTextIds2[CURRENT_DAY - 1]);
+    } else if ((sCsIdList[this->daytimeIndex] <= CS_ID_NONE) || (play->actorCtx.flags & ACTORCTX_FLAG_TELESCOPE_ON)) {
+        // Previously night, turning day, without a cutscene
+        if (play->actorCtx.flags & ACTORCTX_FLAG_TELESCOPE_ON) {
             Sram_IncrementDay();
             gSaveContext.save.time = CLOCK_TIME(6, 0);
-            Message_DisplaySceneTitleCard(play, sDayMessages1[CURRENT_DAY - 1]);
+            Message_DisplaySceneTitleCard(play, sDawnOfTextIds2[CURRENT_DAY - 1]);
         } else {
-            this->csIdIndex = 0;
-            this->unk_146 = gSaveContext.save.time += CLOCK_TIME_MINUTE;
+            this->daytimeIndex = THREEDAY_DAYTIME_NIGHT;
```

Comment:
Do you understand why it's setting `this->daytimeIndex` to night here? Because I don't, and I'm a bit confused as to what this code is trying to accomplish. To get to this block of code, `this->daytimeIndex` already has to be night, right?

Reply:
You're right in that the only way this code is reached is if `this->daytimeIndex` is already set to `THREEDAY_DAYTIME_NIGHT` so this line of code is completely useless.

The only comment I can make about this is that this is being set as part of `EnTest4_HandleDayNightSwapFromInit` and not `EnTest4_HandleDayNightSwap`, which I'm not sure why `EnTest4_HandleDayNightSwapFromInit` exists when there's the near identical `EnTest4_HandleDayNightSwap` which is actually what's used every frame. In practice, `EnTest4_HandleDayNightSwap` is what' always used so that seems like the better designed function.

Reply:
Glad I wasn't mistaken about how strange this code is. It looks like this can only get called in the first place if you manage to call `EnTest4_Init` right at 6:00 AM. And, in fact, in that circumstance, `this->daytimeIndex` is set to `THREEDAY_DAYTIME_NIGHT`, so it's more likely to go down this path.